main.c

```diff
@@ -193,36 +194,200 @@ void denoise(uint8_t *out,
+    for (int y = 0; y < height; y++) {
+        uint8_t *scan_in_line = in + y * stride;
+        uint8_t *scan_out_line = out + y * stride * channels;
+        if (y == 0) {
+            for (int x = 0; x < stride; x ++) {
+                int col_sum = 0;
+                int col_sum_pow = 0;
+                for (int z = -radius; z <= radius; z++) {
+                    uint8_t *sample = in + col_off[z] + x;
+                    col_sum += *sample;
+                    col_sum_pow += *sample * *sample;
+                }
+                col_val[x] = col_sum;
+                col_pow[x] = col_sum_pow;
+            }
+        } else {
+            uint8_t *last_col = in + col_off[y - radius - 1];
+            uint8_t *next_col = in + col_off[y + radius];
+            for (int x = 0; x < stride; x++) {
+                col_val[x] -= last_col[x] - next_col[x];
+                col_pow[x] -= last_col[x] * last_col[x] - next_col[x] * next_col[x];
+            }
+        }
+
+        int prev_sum = 0, prev_sum_pow = 0;
+        for (int z = -radius; z <= radius; z++) {
+            int index = row_off[z];
+            prev_sum += col_val[index];
+            prev_sum_pow += col_pow[index];
+        }
+
+        int mean = prev_sum / window_size;
+        int diff = mean - *scan_in_line;
+        int edge = CLAMP2BYTE(diff);
+        int masked_edge =
+            (edge * *scan_in_line + (256 - edge) * mean) >> 8;
+        int var = (prev_sum_pow - mean * prev_sum) / window_size;
+        int out = masked_edge -
+                  diff * var / (var + smooth_table[*scan_in_line]);
+        *scan_out_line = CLAMP2BYTE(out);
+
+        scan_in_line++, scan_out_line += channels;
+        for (int x = 1; x < width; x++) {
+            int last_row = row_off[x - radius - 1];
+            int next_row = row_off[x + radius];
+
+            prev_sum -= col_val[last_row] - col_val[next_row];
+            prev_sum_pow = prev_sum_pow - col_pow[last_row] +
+                              col_pow[next_row];
+            int mean = prev_sum / window_size;
+            int diff = mean - *scan_in_line;
+            int edge = CLAMP2BYTE(diff);
+            int masked_edge =
+                (edge * *scan_in_line + (256 - edge) * mean) >> 8;
+            int var = (prev_sum_pow - mean * prev_sum) / window_size;
+            int out = masked_edge -
+                      diff * var / (var + smooth_table[*scan_in_line]);
+            scan_out_line[ch_idx] = CLAMP2BYTE(out);
+
+            scan_in_line++, scan_out_line += channels;
+        }
+    }
+
+    free(col_pow);
+    free(col_val);
+    free(row_pos);
+    free(col_pos);
+}
+
 static void die(char *msg)
 {
     fprintf(stderr, "Fatal: %s\n", msg);
     exit(-1);
 }
 
+
+inline uint64_t time_diff(struct timeval *st, struct timeval *et)
+{
+    return (et->tv_sec - st->tv_sec)*1000000ULL + (et->tv_usec - st->tv_usec);
+}
+
 int main(int argc, char *argv[])
 {
-    if (argc < 2)
+    if (argc < 2){
```

Comment:
Please use `clang-format -i *.[ch]` to ensure consistent indentation.

Reply:
For algorithm-based implementation, I don't think format tool give meaningful output.
Splitting to multiple lines or merging to single line has its meaning for reading/thinking or explanation.
Even reduce space between multiplication '*' has its advantage for study.
Of course you can keep it formatted in your repo.
Anyway, I will not apply formatted pull requests.

Reply:
Clang-format understands also special comments that switch formatting in a delimited range. The code between a comment `/* clang-format off */` up to a comment `/* clang-format on */` will not be formatted. See https://clang.llvm.org/docs/ClangFormatStyleOptions.html

Reply:
Yeah, I know that it can be enabled / disabled by comments.
But it is not useful in many cases, just break code readability with low priority rules.
I will show my point with a patch.
Besides, I also want to point out that clang-format causes compilation failed due to incorrect modifications of stb_image.h.